src/frontend/src/optimizer/plan_node/stream_index_scan.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+
+use itertools::Itertools;
+use risingwave_pb::stream_plan::stream_node::Node as ProstStreamNode;
+use risingwave_pb::stream_plan::StreamNode as ProstStreamPlan;
+
+use super::{LogicalScan, PlanBase, PlanNodeId, ToStreamProst};
+use crate::optimizer::property::Distribution;
+
+/// `StreamIndexScan` is a virtual plan node to represent a stream table scan. It will be converted
+/// to chain + merge node (for upstream materialize) + batch table scan when converting to `MView`
+/// creation request. Compared with `StreamTableScan`, it will reorder columns, and the chain node
+/// doesn't allow rearrange.
+#[derive(Debug, Clone)]
+pub struct StreamIndexScan {
+    pub base: PlanBase,
+    logical: LogicalScan,
+    batch_plan_id: PlanNodeId,
+}
+
+impl StreamIndexScan {
+    pub fn new(logical: LogicalScan) -> Self {
+        let ctx = logical.base.ctx.clone();
+
+        let batch_plan_id = ctx.next_plan_node_id();
+        // TODO: derive from input
+        let base = PlanBase::new_stream(
+            ctx,
+            logical.schema().clone(),
+            logical.base.pk_indices.clone(),
+            Distribution::AnyShard, // Mark as `AnyShard` cause we don't know the distribution yet.
+            false,                  // TODO: determine the `append-only` field of table scan
+        );
+        Self {
+            base,
+            logical,
+            batch_plan_id,
+        }
+    }
+
+    pub fn table_name(&self) -> &str {
+        self.logical.table_name()
+    }
+
+    pub fn logical(&self) -> &LogicalScan {
+        &self.logical
+    }
+}
+
+impl_plan_tree_node_for_leaf! { StreamIndexScan }
+
+impl fmt::Display for StreamIndexScan {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "StreamIndexScan {{ index: {}, columns: [{}], pk_indices: {:?} }}",
+            self.logical.table_name(),
+            self.logical.column_names().join(", "),
+            self.base.pk_indices
+        )
+    }
+}
+
+impl ToStreamProst for StreamIndexScan {
+    fn to_stream_prost_body(&self) -> ProstStreamNode {
+        unreachable!("stream index scan cannot be converted into a prost body -- call `adhoc_to_stream_prost` instead.")
+    }
+}
+
+impl StreamIndexScan {
+    pub fn adhoc_to_stream_prost(&self, auto_fields: bool) -> ProstStreamPlan {
```

Comment:
What does the "adhoc" mean here?

Reply:
Generally, `to_stream_prost` only produce body of one stream node. However, StreamIndexScan / StreamTableScan corresponds to 3 stream nodes. Therefore, we have to workaround the current `to_stream` implementation in both plan nodes.

Reply:
Yes, but The term "adhoc" may be a little ambiguous with "ad-hoc query"

Reply:
Then let's rename it to `stream_index_scan_to_stream_prost` and `stream_table_scan_to_stream_prost` later 🤣